### PR TITLE
RavenDB-16517 - Offline activation of a license

### PR DESCRIPTION
**Pull request by grisha-kotler**
https://issues.hibernatingrhinos.com/issue/RavenDB-16517

### CODE REVIEW

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 1]
```diff
@@ -492,7 +492,44 @@ public void TryActivateLicense(bool throwOnActivationFailure)
...
+            var leasedLicenseAsStream = await httpResponseMessage.Content.ReadAsStreamAsync().ConfigureAwait(false);
+            using (var context = JsonOperationContext.ShortTermSingleUse())
+            {
+                var json = context.Read(leasedLicenseAsStream, "leased license info");
+                var leasedLicense = JsonDeserializationServer.LeasedLicense(json);
+                return leasedLicense;
+            }
+        }
+
+        /*public async Task<License> GetUpdatedLicense(
```
Comment (ayende) [inline src/Raven.Server/Commercial/LicenseManager.cs]: ???

File: src/Raven.Server/Commercial/LicenseManager.cs [hunk 2]
```diff
@@ -556,45 +593,89 @@ private StudioConfiguration.StudioEnvironment GetStudioEnvironment()
...
+        {
+            var license = _licenseHelper.TryGetLicenseFromString(throwOnFailure: false) ??
+                          _licenseHelper.TryGetLicenseFromPath(throwOnFailure: false);
+
+            // since we are updating NOT from api.ravendb.net, we are enforcing:
+            // - the same license id (can cause issues when updating the let's encrypt certificate)
+            // - higher expiration date
+
+            if (license == null || license.Id != currentLicense.Id)
+                return null;
```
Comment (ayende) [inline src/Raven.Server/Commercial/LicenseManager.cs:629]: Don't we need to give an error here? The user may be confused on why they are unable to update the license.

File: src/Raven.Server/Commercial/License.cs [hunk 3]
```diff
@@ -22,6 +23,14 @@ public DynamicJsonValue ToJson()
             };
         }
 
+        public override bool Equals(object obj)
```
Comment (ppekrol) [inline src/Raven.Server/Commercial/License.cs:26]:
```
'License' overrides Object.Equals(object o) but does not override Object.GetHashCode()
```

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Commercial/LicenseManager.cs]: Ping

Comment (grisha-kotler) [reply src/Raven.Server/Commercial/LicenseManager.cs]: done

File: src/Raven.Server/Commercial/LicenseManager.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Commercial/LicenseManager.cs:629]: done

File: src/Raven.Server/Commercial/License.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Commercial/License.cs:26]: done

**Review by ayende — APPROVED** [review]

**Pull request merged**